### PR TITLE
split Image `interpolation` on two properties, `interpoltion2d` and `interpolation3d` 

### LINKED ISSUE #3551: 3D interpolation mode cannot be set at layer creation time
## 🐛 Bug

Using `viewer.add_image(image, interpolation='nearest')` when the viewer is in 3D mode doesn't work.

```python
import numpy as np
import napari

image = (np.random.random((5, 5, 5)) > 0.7).astype(int)

viewer = napari.viewer(ndisplay=3)
image_layer = viewer.add_image(image, interpolation='nearest')
# image_layer.interpolation = 'nearest'  # uncomment to get correct interpolation

napari.run()
```

napari info

```
napari: 0.4.12.dev18+gc53f34e85
Platform: macOS-11.4-x86_64-i386-64bit
System: MacOS 11.4
Python: 3.9.7 | packaged by conda-forge | (default, Sep  2 2021, 17:58:46)  [Clang 11.1.0 ]
Qt: 5.15.2
PyQt5: 5.15.4
NumPy: 1.20.3
SciPy: 1.7.1
Dask: 2021.09.0
VisPy: 0.9.3

OpenGL:
  - GL version:  2.1 ATI-4.5.14
  - MAX_TEXTURE_SIZE: 16384

Screens:
  - screen 1: resolution 2304x1296, scale 2.0
  - screen 2: resolution 1792x1120, scale 2.0

Plugins:
  - animation: 0.0.2
  - console: 0.0.4
  - napari-midi: 0.1.dev5+g16c91f6.d20210914
  - napari-ome-zarr: 0.3.0
  - nd2-reader: 0.0.6
  - platelet-unet-watershed: 0.0.3
  - scikit-image: 0.4.12.dev18+gc53f34e85
  - svg: 0.1.5
  - tracks: 0.0.2
  - zarpaint: 0.2.0
```

### PR DESCRIPTION
**Pull request by Czaki**
# Description

This PR simplifies access to interpolation of the Image layer by splitting it into two properties, one for each method of visualization. This allows to set interpolation of 3d mode without changing dimensions to 3d and simplifies the coping layer. 

## Type of change

- [x] Bug-fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# References

closes #3551

# How has this been tested?

- [ ] example: the test suite for my feature covers cases x, y, and z
- [x] example: all tests pass with my change
- [ ] example: I check if my changes works with both PySide and PyQt backends
      as there are small differences between the two Qt bindings.  

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/docs/dev/guides/translations.html).

### CODE REVIEW

File: napari/utils/migrations.py [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+            if from_name in kwargs:
+                warnings.warn(
+                    trans._(
+                        "Argument {from_name} is deprecated, please use {to_name} instead.",
+                        from_name=from_name,
+                        to_name=to_name,
+                    ),
+                    category=FutureWarning,
+                    stacklevel=2,
+                )
```
Comment (Carreau) [inline napari/utils/migrations.py]: If you say "deprecated", then likely the category should be "DeprecationWarning". 
Also you most likely want to include _at least_, the version since when so that people have an indication as to whether they can/should update depending on the their requirements.

You also may want to check that not both  old and new are used at the same time.

File: napari/layers/image/image.py [hunk 2]
```diff
@@ -292,6 +295,8 @@ def __init__(
         self.events.add(
             mode=Event,
             interpolation=Event,
```
Comment (Carreau) [inline napari/layers/image/image.py]: Should this event also be deprecated ?

File: napari/layers/image/image.py [hunk 3]
```diff
@@ -527,18 +527,46 @@ def interpolation(self):
...
+        )
 
     @interpolation.setter
     def interpolation(self, interpolation):
         """Set current interpolation mode."""
+        warnings.warn(
+            trans._(
+                "Interpolation setting is deprecated. Please use interpolation2d or interpolation3d",
+            ),
+            category=FutureWarning,
```
Comment (Carreau) [inline napari/layers/image/image.py]: Same comment, if you say "deprecated", I guess the category should be  `DeprecationWarning` and the message should say since when.

```suggestion
            category=DeprecationWarning,
```

Also thanks for stacklevel !

Comment (Czaki) [reply napari/layers/image/image.py]: The problem with deprecation Warning is that they are not visible during the normal run of python code. So if the plugin creator does not test properly his code against deprecation he will do not know that he use deprecated code https://docs.python.org/3/library/exceptions.html#DeprecationWarning

File: napari/utils/migrations.py (hunk 1 above)
Comment (Czaki) [reply napari/utils/migrations.py]: tested, and deprecation version added.

File: napari/layers/image/image.py (hunk 2 above)
Comment (Czaki) [reply napari/layers/image/image.py]: Done

**Comment by brisvag** [comment]
This is a reasonable, but we should discuss a bit if we want it. In the past, we had discussion about splitting or unifying the interpolation between 2D and 3D in #4467.

In short, with the next vispy release we will be able to have a single interpolation, which gets used for both, because interpolations will be the same set for 2D and 3D; thanks to this, we wouldn't need this change.

On the other hand using anything with a bigger kernel than linear in 3D can kill the performance *really* fast, which would be a problem if someone wants to switch from a highly interpolated 2D to 3D view, and suddenly evenerything freezes. In light of this, I think that splitting 2D and 3D is still a good idea, despite the fact that we don't necessarily *need* it. What do you think @andy-sweet ?

**Comment by Czaki** [comment]
> In short, with the next vispy release we will be able to have a single interpolation, which gets used for both, because interpolations will be the same set for 2D and 3D; thanks to this, we wouldn't need this change.

For me, this is not a full valid argument. As interpolation in 2d and in 3d has different semantic meanings. For example, current behavior when 2d is done using nearest and 3d is done using linear is a good example of why keeping this 2 properties is a good approach.

File: napari/layers/image/image.py (hunk 3 above)
Comment (Carreau) [reply napari/layers/image/image.py]: Hum I see your point, but I think if we want to make those warnings visible in napari, we  need to insert a `DeprecationWarning` filter to default at napari startup, potentially with a setting option to disable that. We can limit it to `module='napari.*'` to only target our modules.

Comment (Czaki) [reply napari/layers/image/image.py]: I see your point, but current strategy in napari codebase is to use FutureWarning.

https://github.com/napari/napari/blob/9cfb915e01a69e9717fad193533d0a6bcb92d98f/napari/qt/progress.py#L5-L12

https://github.com/napari/napari/blob/d490354431bf265d580e73f15b7bfebdfb45e48e/napari/_qt/qt_event_loop.py#L265-L271

I like the idea of your change, but maybe it is worth opening a separate PR/issue for this to allow discussion in the wider group?

Comment (Carreau) [reply napari/layers/image/image.py]: opened, https://github.com/napari/napari/issues/4731

> I see your point, but current strategy in napari codebase is to use FutureWarning.

But at the same time there is half a dozen DeprecationWarnings already emitted in napari/layers/utils/text_manager.py and a few others in napari/layers/base/base.py, napari/layers/utils/layer_utils.py , napari/resources/__init__.py


I think in total there are more Deprecation Warnings than FutureWarning. In any case I agree that we should be consistent.

Comment (Czaki) [reply napari/layers/image/image.py]: Done

File: napari/layers/image/image.py [hunk 4]
```diff
@@ -527,18 +534,46 @@ def interpolation(self):
         str
             The current interpolation mode
         """
-        return str(self._interpolation[self._ndisplay])
+        return str(
+            self._interpolation2d
+            if self._ndisplay == 2
+            else self._interpolation3d
+        )
```
Comment (brisvag) [inline napari/layers/image/image.py:548]: Either this should be deprecated too, or we should keep the setter. I'm fine either way, but maybe it would be less confusing to simply remove it.

File: napari/utils/migrations.py [hunk 5]
```diff
@@ -0,0 +1,48 @@
+import warnings
+from functools import wraps
+
+from napari.utils.translations import trans
+
+
+def rename_argument(from_name: str, to_name: str, version: str):
```
Comment (brisvag) [inline napari/utils/migrations.py:7]: Cool!

**Review by brisvag — COMMENTED** [review]
I'm not super conviced about having two separate attributes... A nested model would be cool (`image.interpolation.2d = ...`), but without something like #4474 we'd have to hook up events ourselves I think.

File: napari/layers/image/image.py [hunk 6]
```diff
@@ -354,15 +366,10 @@ def __init__(
...
-            2: Interpolation.NEAREST,
-            3: (
-                Interpolation3D.NEAREST
-                if self.__class__.__name__ == 'Labels'
-                else Interpolation3D.LINEAR
-            ),
-        }
-        self.interpolation = interpolation
+        self._interpolation2d = Interpolation.NEAREST
+        self._interpolation3d = Interpolation3D.NEAREST
```
Comment (andy-sweet) [inline napari/layers/image/image.py:370]: Why do you need to set the private attributes here? Won't the setters called immediately below take care of that?

**Comment by Czaki** [comment]
> A nested model would be cool (`image.interpolation.2d = ...`),

we cannot have attribute named `2d`. `3d` is also prohibited as both start with digit. 

In the longer perspective we should split data from representation.

File: napari/layers/image/image.py (hunk 4 above)
Comment (andy-sweet) [reply napari/layers/image/image.py:548]: Not sure if you added/updated the setter since this comment, but if the setter is deprecated (and it probably should be), my opinion is that the getter should be too.

File: napari/layers/image/image.py [hunk 7]
```diff
@@ -1016,7 +1051,8 @@ def _get_state(self):
                 'multiscale': self.multiscale,
                 'colormap': self.colormap.name,
                 'contrast_limits': self.contrast_limits,
-                'interpolation': self.interpolation,
```
Comment (andy-sweet) [inline napari/layers/image/image.py:1019]: This will break plugins that used the `interpolation` key when receiving a layer data tuple. My guess is that's not many plugins (if any at all) and the fix is pretty simple, so I'm OK with removing this, but just wanted to flag.

File: napari/utils/migrations.py [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+    to_name : str
+        new name of argument
+    """
+
+    def _wrapper(func):
+        @wraps(func)
+        def _update_from_dict(*args, **kwargs):
+            if from_name in kwargs:
+                if to_name in kwargs:
+                    raise ValueError(
```
Comment (andy-sweet) [inline napari/utils/migrations.py:25]: I had not thought to do this in previous deprecations, but I like this logic.

**Review by andy-sweet — CHANGES_REQUESTED** [review]
I'm happy with splitting these into two attributes as long as we maintain them that way for some amount of time. Therefore, this PR looks good to me except for a couple of minor issues (see line comments).

File: napari/layers/image/image.py (hunk 6 above)
Comment (Czaki) [reply napari/layers/image/image.py:370]: there is strong advise to not define attribute outside constructor.

File: napari/utils/migrations.py (hunk 5 above)
Comment (andy-sweet) [reply napari/utils/migrations.py:7]: Agreed! Very nice!

File: napari/layers/image/image.py (hunk 3 above)
Comment (andy-sweet) [reply napari/layers/image/image.py]: I was even aware of `FutureWarning`, so thanks for the info, and thanks for #4731 too!

File: napari/layers/image/image.py (hunk 7 above)
Comment (Czaki) [reply napari/layers/image/image.py:1019]: fix is not so easy as dis dict need to match constructor arguments. need to pop `interpolation` before pass to constructor is why I do not add it. 

but maybe @tlambert03 has other opinion/experience?

**Comment by brisvag** [comment]
> we cannot have attribute named `2d`. `3d` is also prohibited as both start with digit.

Ah, right.

File: napari/layers/image/image.py (hunk 7 above)
Comment (brisvag) [reply napari/layers/image/image.py:1019]: Yeah, this is a recurring issue with deprecations :/

File: napari/layers/image/image.py (hunk 6 above)
Comment (andy-sweet) [reply napari/layers/image/image.py:370]: Oh right. I think that I would personally break that rule here because it's pretty clear what's going on (i.e. we're using the corresponding property setters and not some other method).

But I don't feel strongly, so feel free to leave this as is.

File: napari/layers/image/image.py (hunk 7 above)
Comment (andy-sweet) [reply napari/layers/image/image.py:1019]: Yeah, I'm aware of that problem and it irks me. To clarify, I definitely don't expect a fix in this PR and I think it's OK to break things here.

File: napari/layers/image/image.py [hunk 9]
```diff
@@ -527,18 +534,46 @@ def interpolation(self):
...
+            ),
+            category=DeprecationWarning,
+            stacklevel=2,
+        )
         if self._ndisplay == 3:
-            self._interpolation[self._ndisplay] = Interpolation3D(
-                interpolation
-            )
+            self.interpolation3d = interpolation
+            self.events.interpolation(value=self.interpolation3d)
```
Comment (andy-sweet) [inline napari/layers/image/image.py]: Instead of emitting the deprecated `interpolation` event here, I would expect it to be emitted in the `interpolation2d` and `interpolation3d` setters instead. That's because the event should indicate that the value I'll get from `Image.interpolation` could have changed, which is the case if the non-deprecated `interpolation2d` and `interpolation3d` setters are used.

File: napari/layers/image/image.py [hunk 10]
```diff
@@ -527,18 +534,46 @@ def interpolation(self):
...
         """
-        return str(self._interpolation[self._ndisplay])
+        return str(
+            self._interpolation2d
+            if self._ndisplay == 2
+            else self._interpolation3d
+        )
 
     @interpolation.setter
     def interpolation(self, interpolation):
```
Comment (andy-sweet) [inline napari/layers/image/image.py:551]: `QtImageControls.changeInterpolation` is still using this deprecated setter, but GitHub won't let me comment there. That means that using the GUI control to change the value will always raise a warning, right? I think we need a fix, which isn't just `filterwarnings`.

Comment (Czaki) [reply napari/layers/image/image.py:551]: I do not understand why I do not notice warnings when running test.

File: napari/layers/image/image.py (hunk 9 above)
Comment (Czaki) [reply napari/layers/image/image.py]: done

File: napari/layers/image/image.py (hunk 4 above)
Comment (Czaki) [reply napari/layers/image/image.py:548]: done

**Comment by andy-sweet** [comment]
Will re-review, but looks like there are a few check failures that are probably not coincidental.

File: tox.ini [hunk 11]
```diff
@@ -80,9 +80,9 @@ commands_pre =
     # the rest is for good measure
     headless: pip uninstall -y pytest-qt qtpy pyqt5 pyside2
 commands =
-    !headless: python -m pytest {env:PYTEST_PATH:} --color=yes --basetemp={envtmpdir} \
+    !headless: python -m pytest . --color=yes --basetemp={envtmpdir} \
         --cov-report=xml --cov={env:PYTEST_PATH:napari} --ignore tools --maxfail=5 \
-        --json-report --json-report-file={toxinidir}/report-{envname}.json {posargs}
+        --json-report --json-report-file={toxinidir}/report-{envname}.json {posargs} \
```
Comment (andy-sweet) [inline tox.ini]: Do you need to try to fix this file in this PR? Seems like that's causing the check failures.

**Review by andy-sweet — APPROVED** [review]
Looks good to me, though maybe would be get an approval from @brisvag too.

File: tox.ini [hunk 12]
```diff
@@ -80,10 +80,10 @@ commands_pre =
     # the rest is for good measure
     headless: pip uninstall -y pytest-qt qtpy pyqt5 pyside2
 commands =
-    !headless: python -m pytest {env:PYTEST_PATH:} --color=yes --basetemp={envtmpdir} \
+    !headless: python -m pytest napari --color=yes --basetemp={envtmpdir} \
         --cov-report=xml --cov={env:PYTEST_PATH:napari} --ignore tools --maxfail=5 \
-        --json-report --json-report-file={toxinidir}/report-{envname}.json {posargs}
-        --skip_examples
+        --json-report --json-report-file={toxinidir}/report-{envname}.json \
+        --skip_examples {posargs}
```
Comment (brisvag) [inline tox.ini:86]: Why are these changes here?

Comment (Czaki) [reply tox.ini:86]: because `--skip_examples` was ignored and the test executes also examples tests. This makes tests with tox fails local for me. 

I also need this for #4665

File: napari/_qt/layer_controls/qt_image_controls.py [hunk 13]
```diff
@@ -52,10 +57,17 @@ class QtImageControls(QtBaseImageControls):
         Label for the rendering mode dropdown menu.
     """
 
+    layer: 'napari.layers.Image'
```
Comment (tlambert03) [inline napari/_qt/layer_controls/qt_image_controls.py:60]: love the type hints thank you.  Can you put all of these type hints on the parameter itself though?  There's a subtle difference, which is that when you do it like this there's no error if you pass in the wrong type:

Screen Shot 2022-06-26 at 6 33 12 AM

Screen Shot 2022-06-26 at 6 33 38 AM

**Review by tlambert03 — APPROVED** [review]

File: napari/_qt/layer_controls/qt_image_controls.py (hunk 13 above)
Comment (Czaki) [reply napari/_qt/layer_controls/qt_image_controls.py:60]: I could put it in the constructor signature too, but cannot remove this without playing with generics. Because in your example there is a direct assignment to a variable and in this code part the assignment is done in the super class constructor. So annotating only constructor parameter will not help with finding all usage that is required during code refactoring. 

Especially, in this case constructor parameter is annotated in docstring, so PyCharm will highlight that the wrong type passed, but still will not know that the `.layer` attribute is the `Image` type.

Comment (tlambert03) [reply napari/_qt/layer_controls/qt_image_controls.py:60]: ok fair enough :)  (I don't mind generics fwiw, but agreed it's a lot more)

**Review by brisvag — APPROVED** [review]

**Comment by brisvag** [comment]
Sounds like we can merge!

File: tox.ini (hunk 12 above)
Comment (tlambert03) [reply tox.ini:86]: Hey @Czaki, I noticed that i'm unable to use posargs anymore to restrict tox tests and tracked it down to this PR.  If I opened a PR to restore the `{env:PYTEST_PATH:}`... is it going to break something for you?  I'd like to be able to run tox without having to run the whole test suite but this breaks that
